manager/state/raft/raft.go

```diff
@@ -810,25 +811,67 @@ func (n *Node) ResolveAddress(ctx context.Context, msg *api.ResolveAddressReques
 	return &api.ResolveAddressResponse{Addr: member.Addr}, nil
 }
 
-// LeaderAddr returns address of current cluster leader.
-// With this method Node satisfies raftpicker.AddrSelector interface.
-func (n *Node) LeaderAddr() (string, error) {
-	ctx, cancel := context.WithTimeout(n.Ctx, 10*time.Second)
+func checkHealth(conn *grpc.ClientConn) error {
+	client := api.NewHealthClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	if err := WaitForLeader(ctx, n); err != nil {
-		return "", ErrNoClusterLeader
+	resp, err := client.Check(ctx, &api.HealthCheckRequest{Service: "Raft"})
+	if err != nil {
+		return ErrHealthCheckFailure
 	}
-	n.stopMu.RLock()
-	defer n.stopMu.RUnlock()
-	if !n.IsMember() {
-		return "", ErrNoRaftMember
+	if resp.Status != api.HealthCheckResponse_SERVING {
+		return ErrHealthCheckFailure
+	}
+	return nil
+}
+
+func (n *Node) getLeaderConn() (*grpc.ClientConn, error) {
+	leader := n.Leader()
+
+	if leader == n.Config.ID {
+		return nil, raftselector.ErrIsLeader
 	}
-	ms := n.cluster.Members()
-	l := ms[n.leader()]
+	l := n.cluster.Members()[leader]
 	if l == nil {
-		return "", ErrNoClusterLeader
+		return nil, fmt.Errorf("no leader found")
+	}
+	if l.Conn == nil {
+		return nil, raftselector.ErrIsLeader
+	}
+	if checkHealth(l.Conn) != nil {
+		return nil, fmt.Errorf("cluster is unhealthy")
+	}
```

Comment:
Let's not do this health check every time `LeaderConn` is called. It means each proxied requests needs two round-trips: one to check the health of the leader, and another to actually invoke the RPC.

If we keep the health check, we should rate-limit it, perhaps using `golang.org/x/time/rate`.

But I'm not sure I understand why it's necessary. It seems like a best-effort check, since the leader can become unhealthy between the health check and trying to use the connection.